runtime/compiler/optimizer/InlinerTempForJ9.cpp

```diff
@@ -866,80 +831,155 @@ TR_J9InlinerPolicy::createAnchorNodesForUnsafeGetPut(TR::TreeTop* treeTop,
-                                       TR::TreeTop* lowTagCmpTree,
+                                       TR::TreeTop* arrayDirectAccessTreeTop,
+                                       TR::TreeTop* indirectAccessTreeTop,
                                        bool needNullCheck, bool isUnsafeGet,
                                        bool conversionNeeded,
-                                       TR::Block * joinBlock,
-                                       TR_OpaqueClassBlock *javaLangClass,
                                        TR::Node* orderedCallNode = NULL)
    {
    TR::CFG *cfg = comp()->getFlowGraph();
-   TR::Block *nullComparisonBlock = prevTreeTop->getEnclosingBlock();
-   TR::TreeTop* nullComparisonTree = nullComparisonBlock->getLastRealTreeTop();
-   TR::TreeTop *nullComparisonEntryTree = nullComparisonBlock->getEntry();
-   TR::TreeTop *nullComparisonExitTree = nullComparisonBlock->getExit();
-   //if conversionNeeded is true, we haven't generated and we don't need arrayDirectAccessBlock
-   TR::Block *arrayDirectAccessBlock = conversionNeeded ? nullComparisonTree->getNode()->getBranchDestination()->getNode()->getBlock() : NULL;
+
+   // Generate the tree for the null comparison
+   TR::Node *addrLoad =
+      TR::Node::createWithSymRef(unsafeAddress,
+                      comp()->il.opCodeForDirectLoad(unsafeAddress->getDataType()),
+                      0, newSymbolReferenceForAddress);
+   TR::Node *nullComparisonNode =
+      TR::Node::createif(TR::ifacmpeq, addrLoad,
+                        TR::Node::create(addrLoad, TR::aconst, 0, 0), NULL);
+   TR::TreeTop *nullComparisonTree = TR::TreeTop::create(comp(), nullComparisonNode);
+
+   // If low-order bit of offset is set, perform indirect access.  The IL for the case
+   // that handles conversion places the indirect access code on the fall-through
+   // path, so it needs to branch if the bit is not set; otherise, the indirect
+   // access code is on the taken path of the branch, so it needs to branch if the
+   // bit is set.
+   //
+   TR::TreeTop *lowTagCmpTree = genClassCheckForUnsafeGetPut(unsafeOffset, /* branchIfLowTagged */ !conversionNeeded);
+
+   TR::TreeTop *firstComparisonTree;
+   TR::TreeTop *ifTree;
+   TR::TreeTop *elseTree;
+
+   TR_OpaqueClassBlock *javaLangClass = comp()->getClassClassPointer(/* isVettedForAOT = */ true);
+
+   // Determine overall layout of IL - which test to place first (null test of object or
+   // setting of low-order bit of offset) - and whether direct or indirect access IL will
+   // be placed on the fall-through path of the first branch.
+   //
+   if (conversionNeeded)
+      {
+      firstComparisonTree = nullComparisonTree;
+      ifTree = arrayDirectAccessTreeTop;
+      elseTree = indirectAccessTreeTop;
+      }
+   else
+      {
+      firstComparisonTree = (javaLangClass != NULL) ? lowTagCmpTree : nullComparisonTree;
+      ifTree = indirectAccessTreeTop;
```

Comment:
Why would we want the "indirect access tree" (which I assume is the access via ram statics) as the fall through in the case when we are doing a "no conversion" access ? i.e. int or long. I feel accessing statics via unsafe is less common in general, and so maybe we should consider swapping this (can be in a future PR, if its too much work to change in this one). I am also not sure if somehow this order gets fixed up later in the below implementation, but I am adding this comment as I review the code.

Reply:
I think there might be some confusion about the terminology I've used, which came from the arguments and implementation of [`OMR::Block::createConditionalBlocksBeforeTree`](https://github.com/eclipse/omr/blob/4856f9da7c1f6a02e2c25308a4afcee6ceb0694d/compiler/il/OMRBlock.cpp#L1898-L1918).  In that case, the `ifTree` is the target or taken-side of the branch, while `elseTree` is the fall-through branch.  So in this case, it falls-through to the "direct access tree".

Perhaps I should change the names of `ifTree` and `elseTree` to `targetOfIfTree` and `fallThroughTree`, or something like that, to keep it clear how these trees will be arranged.  Thoughts?

Reply:
Sure that naming would make it clearer. 

Your clarification of the `ifTree` meaning helps, though it does raise a question about the other side of this if-then-else in this code where the `indirectAccessTreeTop` is then the fall though tree, essentially (and the array direct access tree is the branch target). Maybe that is not a case we were intending to change any behavior for in this PR, in which case, we don't have to change it in this PR (i.e. leave for later). But, in general, it seems like having the indirect access tree as the fall through may not be the best idea, since that is never expected to be the common case.

Reply:
> Maybe that is not a case we were intending to change any behavior for in this PR, in which case, we don't have to change it in this PR (i.e. leave for later). But, in general, it seems like having the indirect access tree as the fall through may not be the best idea, since that is never expected to be the common case.

Yes, as you suggest, that was the original behaviour for the conversion case, and I hadn't wanted to alter that for this pull request.  I do have a prototype change that rearranges things so that the "direct access" code lies on the fall-through path for all three cases that I can work on as a follow-on change.

Reply:
Ok that sounds good, thanks Henry. You can plan to deliver those changes afterwards.

Reply:
Revised naming of "ifTree" and "elseTree" in commit 0be738165c